Revise “Container runtimes” page in light of dockershim removal

PranshuSrivastava opened the pull request:
This updates the container-runtimes [page](https://kubernetes.io/docs/setup/production-environment/container-runtimes/ of the website to include information about the Dockershim deprecation and so that the users can take an informed decision. This is in accordance with #28449

PranshuSrivastava commented:
/cc @sftim

PranshuSrivastava commented:
@sftim I had to keep the links this way because they were not being rendered properly if I broke them in between.

bradtopol commented:
Hi @mikebrow Can you serve as the technical reviewer for this PR? Just add a /lgtm if it is technically correct and I'll take care of the rest.  Thanks. 
/cc @mikebrow 
/assign

mikebrow reviewed (approved):
/lgtm

sftim commented:
/label tide/merge-method-squash

I'd prefer to omit the whitespace changes, but it's going to be easier I think to merge this and then do fixups.

PranshuSrivastava commented:
> /label tide/merge-method-squash
> 
> I'd prefer to omit the whitespace changes, but it's going to be easier I think to merge this and then do fixups.

@sftim those whitespace changes are because of my text editor.

chrisnegus commented:
/assign @celestehorgan

reylejano commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -22,6 +22,17 @@ Kubernetes, on Linux:
 - [Docker](#docker)
 
+{{< note >}}
+Docker container runtime (Dockershim) has been [deprecated](
```
This implies that Docker container runtime is deprecated and that dockershim and Docker container runtime are the same.
I suggest more clarity that dockershim is deprecated

sftim replied:
@PranshuSrivastava would you be willing to revise the language here?

PranshuSrivastava replied:
@sftim @reylejano sorry I missed this review. I have made the change, PTAL.

sftim commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -22,6 +22,17 @@ Kubernetes, on Linux:
 - [Docker](#docker)
 
+{{< note >}}
+Dockershim, the module that allowed Kubernetes to interact with Docker has been [deprecated](
```
Rather than “has been”, I recommend “was”. The reason is that “has been” suggests it's a recent removal and we prefer the text of documentation to be more timeless (and easier to maintain).

PranshuSrivastava replied:
@mikebrow  @sftim done.

mikebrow reviewed (approved):
LGTM

PranshuSrivastava commented:
@vincepri @bart0sh PTAL.

bart0sh commented:
@PranshuSrivastava please, squash your commits. otherwise lgtm.

PranshuSrivastava commented:
> Hi @PranshuSrivastava, it looks like due to a merge conflict, this PR cannot be merged as of now.
> 
> I would highly suggested you to perform a [rebase](https://kubernetes.io/docs/contribute/new-content/open-a-pr/#merge-conflicts-and-rebasing) (it is a great skill to learn in any source control) on this PR after which you can [squash the commits.](https://kubernetes.io/docs/contribute/new-content/open-a-pr/#merge-conflicts-and-rebasing)
> 
> If you need a hand/help in doing any of this, feel free to ask your questions in the [SIG-docs Slack.](https://kubernetes.slack.com/messages/sig-docs)
> 
> Also if you are not comfortable in doing a _rebase_ of this PR please do tell, in that case any Docs contributor will tell you the way forward .
> 
> Thanks.

@Babapool Done, thanks.

PranshuSrivastava commented:
> @PranshuSrivastava please, squash your commits. otherwise lgtm.

@bart0sh done

tengqm commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -29,6 +29,21 @@ Kubernetes, on Linux:
 
+{{< note >}}
+Docker container runtime (Dockershim) has been [deprecated](
+https://kubernetes.io/blog/2020/12/02/dockershim-faq/)
```
```suggestion
/blog/2020/12/02/dockershim-faq/)
```

tengqm commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -29,6 +29,21 @@ Kubernetes, on Linux:
+Docker container runtime (Dockershim) has been [deprecated](
+https://kubernetes.io/blog/2020/12/02/dockershim-faq/)
+as of [version 1.20](
+https://kubernetes.io/blog/2020/12/08/kubernetes-1-20-release-announcement/#dockershim-deprecation)
```
We use relative paths rather than absolute ones. Also please help make sure the link appears on one line.
```suggestion
/blog/2020/12/08/kubernetes-1-20-release-announcement/#dockershim-deprecation)
```

tengqm commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -29,6 +29,21 @@ Kubernetes, on Linux:
+and is planned to be removed in by 1.24, therefore to use Docker you need to
+be running an older version of Kubernetes.
+You can check out this [documentation](
+https://kubernetes.io/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you/)
```
```suggestion
/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you/)
```

tengqm commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -29,6 +29,21 @@ Kubernetes, on Linux:
+You can check out this [documentation](
+https://kubernetes.io/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you/)
+to understand how this deprecation might affect you. For migrating from
+dockershim you can follow [this](
+https://kubernetes.io/docs/tasks/administer-cluster/migrating-from-dockershim/)
```
```suggestion
dockershim you can follow [the migration guide](/docs/tasks/administer-cluster/migrating-from-dockershim/)
```

reylejano replied:
@PranshuSrivastava check the links from the preview, @tengqm 's suggestion is correct

PranshuSrivastava replied:
@reylejano Okay

tengqm commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -29,6 +29,21 @@ Kubernetes, on Linux:
+to understand how this deprecation might affect you. For migrating from
+dockershim you can follow [this](
+https://kubernetes.io/docs/tasks/administer-cluster/migrating-from-dockershim/)
+document to migrate from dockershim.
```
```suggestion
to migrate from dockershim.
```

tengqm commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -416,10 +431,10 @@ in sync.
-this container runtime is available in various forms.
-[Install Docker Engine](https://docs.docker.com/engine/install/) explains your options
-for installing this runtime.
+1. On each of your nodes, install the Docker for your Linux distribution as per
```
```suggestion
On each of your nodes, install the Docker for your Linux distribution as per
```

tengqm commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -428,7 +443,10 @@ You can also find third-party adapters that let you use Docker Engine with Kuber
 (CRI).
 
-The following CRI adaptors are designed to work with Docker Engine:
+   {{< note >}}
+   `overlay2` is the preferred storage driver for systems running Linux kernel version 4.0 or higher,
+   or RHEL or CentOS using version 3.10.0-514 and above.
+   {{< /note >}}
```
```suggestion
{{< note >}}
`overlay2` is the preferred storage driver for systems running Linux kernel version 4.0 or higher,
or RHEL or CentOS using version 3.10.0-514 and above.
{{< /note >}}
```

PranshuSrivastava commented:
Done @tengqm

reylejano commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -29,6 +29,21 @@ Kubernetes, on Linux:
 - [Mirantis Container Runtime](#mcr)
 
+{{< note >}}
+Docker container runtime (Dockershim) has been [deprecated](
```
```suggestion
Docker container runtime (Dockershim) has been [deprecated](
```
Implies that dockershim and Docker container runtime are the same, please revise

reylejano commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -29,6 +29,21 @@ Kubernetes, on Linux:
+/blog/2020/12/02/dockershim-faq/)
+as of [version 1.20](
+/blog/2020/12/08/kubernetes-1-20-release-announcement/#dockershim-deprecation)
+and is planned to be removed in by 1.24, therefore to use Docker you need to
+be running an older version of Kubernetes.
```
```suggestion
and is planned to be removed in by 1.24, therefore to use Docker you need to
be running an older version of Kubernetes.
```
This is not true, you can use cri-dockerd and use Docker

sftim commented:
/remove-label tide/merge-method-squash

sftim commented:
Thanks for working on this @PranshuSrivastava 

It's an important topic so we're keen to get the nuances and technical details exactly right.

sftim commented:
/sig node
/sig cluster-lifecycle

PranshuSrivastava commented:
@reylejano @sftim PTAL

reylejano commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -416,10 +429,10 @@ in sync.
-this container runtime is available in various forms.
-[Install Docker Engine](https://docs.docker.com/engine/install/) explains your options
-for installing this runtime.
+On each of your nodes, install the Docker for your Linux distribution as per
```
```suggestion
On each of your nodes, install Docker for your Linux distribution as per
```

PranshuSrivastava replied:
done @reylejano

PranshuSrivastava commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -151,9 +164,9 @@ Install containerd:
+1. Install the `containerd.io` package from the official Docker repositories.
+Instructions for setting up the Docker repository for your respective Linux distribution and
+installing the `containerd.io` package can be found at
 [Install Docker Engine](https://docs.docker.com/engine/install/#server).
```
@reylejano PTAL

mikebrow replied:
you can use the containerd release tar provided at containerd.io/downloads.. but that link does not come with an installer or install instructions... more would be needed or linked to in the docs to make that make sense/work (what to do after extracting from the tar for example)

afbjorklund replied:
It's the same story as cri-dockerd, except that doesn't even have downloads...
(I will pretend that the cri-o instructions are easy to follow and set an example.)

* https://github.com/kubernetes/kubernetes/issues/106917#issuecomment-1023964814

afbjorklund replied:
I used the normal k8s packages, that installed crictl 1.19 for me

If you want the latest/greatest, you can go for the ones at Kubic

afbjorklund replied:
@reylejano here are the package dependencies:

Depends: kubelet (>= 1.19.0), kubectl (>= 1.19.0), kubernetes-cni (>= 0.8.7), **cri-tools (>= 1.19.0)**

kubeadm/kubernetes-xenial 1.23.3-00 amd64
kubelet/kubernetes-xenial 1.23.3-00 amd64
kubectl/kubernetes-xenial 1.23.3-00 amd64
kubernetes-cni/kubernetes-xenial 0.8.7-00 amd64
cri-tools/kubernetes-xenial 1.19.0-00 amd64

https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/#installing-kubeadm-kubelet-and-kubectl

You still need to configure /etc/crictl.yaml manually

And to make sure to add a CNI plugin (I used flannel)

sftim commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -29,6 +29,19 @@ Kubernetes, on Linux:
 - [Mirantis Container Runtime](#mcr)
 
+{{< note >}}
+Dockershim, the portion of code in Kubernetes that provided direct
+integration with Docker in prior releases, was removed from Kubernetes
+version 1.24. This removal was announced as a [deprecation in Kubernetes v 1.20](
+/blog/2020/12/08/kubernetes-1-20-release-announcement/#dockershim-deprecation)
+You can check out this [documentation](
+/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you/)
+to understand how this deprecation might affect you. For migrating from
+dockershim you can follow [this migration guide](
+/docs/tasks/administer-cluster/migrating-from-dockershim/)
+to migrate from dockershim.
```
If this text is present, the PR must target the _dev-1.24_ branch and not _main_.

/hold
until that discrepancy is dealt with.

PranshuSrivastava replied:
@sftim done

sftim replied:
Thanks
/unhold

however: we may well want a similar-but-different change targeting _main_, so that we can blog about this before the release of v1.24

This PR can still go ahead, I think.

afbjorklund commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -416,10 +430,10 @@ in sync.
-for installing this runtime.
+On each of your nodes, install Docker for your Linux distribution as per
+[Install Docker Engine](https://docs.docker.com/engine/install/#server).
+You can find the latest validated version of Docker in this
```
I don't think Kubernetes validates against Docker anymore, and there doesn't seem to be a version in that file.

afbjorklund commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -151,10 +164,11 @@ Install containerd:
-Instructions for setting up the Docker repository for your respective Linux distribution and 
-installing the `containerd.io` package can be found at 
-[Install Docker Engine](https://docs.docker.com/engine/install/#server).
+1. Install the `containerd.io` package from the [official containerd website](
```
There are no "containerd.io" packages on the containerd.io website, there it just called "containerd" (and no packages)

The alternative (to "containerd") would be "cri-containerd-cni" which also includes the CRI and CNI configuration:

`/etc/crictl.yaml`

`etc/cni/net.d/10-containerd-net.conflist`

sftim replied:
Where does the package name `cri-containerd-cni` come from?

bart0sh replied:
@sftim cri-containerd-cni is a tarball, recommended as a way to install by the containerd documentation: https://containerd.io/docs/getting-started/#starting-containerd

+1 for not mentioning containerd.io package.

neolit123 replied:
makes sense to link to https://containerd.io/downloads/ and https://containerd.io/docs/getting-started/#starting-containerd

bart0sh replied:
@afbjorklund installing CNI can cause [hard to debug issues](https://github.com/cri-o/cri-o/issues/2885). I'd recommend to install containerd from `cri-containerd` tarball instead.

afbjorklund replied:
Kubernetes only works with one CNI at a time, installing podman or cri-o or even containerd is known to interfere with cluster setup...

The recommendation is to wipe /etc/cni before configuring and starting the cluster

bart0sh replied:
@afbjorklund  So, what's the point to install CNI if it's going to be removed anyway?

btw, I don't see any recommendation to remove CNI in [this documentation](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/). Can you point me out to the recommendation you've mentioned?

afbjorklund replied:
I don't think there is any instructions regarding CNI at all ? So it was mostly how I was adding myself.

`rm -f /etc/cni/net.d/*.conf*`

`kubectl apply -f https://raw.githubusercontent.com/flannel-io/flannel/v0.14.0/Documentation/kube-flannel.yml`

This of course also _breaks_ all other usage, so you need to move the podman configuration directory.

```
# Path to the directory where CNI configuration files are located.
#
# network_config_dir = "/etc/cni/net.d/"
```

Similar with the nerdctl configuration, so that it doesn't interfer with Kubernetes' containerd config (of CNI) ?

afbjorklund replied:
@bart0sh I am not sure what the best way is to package this.

The user _needs_ to set up /etc/crictl.yaml and /etc/cni/net.d

But at the same time, there can be only one container runtime and only one container networking for Kubernetes.

So the files from one will conflict with the files from the other ones. Maybe they should go in /usr/share or in docs ?

bart0sh replied:
@afbjorklund installing https://github.com/containerd/containerd/releases/download/v1.6.0/cri-containerd-1.6.0-linux-arm64.tar.gz and then following [this guide](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/) is the most straightforward way to get working kubernetes cluster in my opinion.

afbjorklund replied:
Sure, but it doesn't have any details ? It says you need a CRI and you need a CNI, but after that it is just "GL;HF".

The entire process was more like: https://github.com/lima-vm/lima/blob/v0.8.3/examples/k8s.yaml#L28_L79

bart0sh replied:
[this guide](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/) was referencing [container runtimes](https://kubernetes.io/docs/setup/production-environment/container-runtimes/) previously IIRC. So, I always followed first guide and it worked for me.

afbjorklund replied:
Minor packaging detail: runc is bundled with "containerd.io" and with "cri-containerd", there is no "runc.io" package.

There is a regular system "runc" package, that is a dependency of the regular system "containerd" package though.

The docker packages are basically internal to docker, they could change (and have changed before) at any time.

Haven't looked at the rpm packaging (this was deb), but assume there is a similar vendor vs. system duplication ?

sftim replied:
For SIG Docs: we want to be as generic as we can be about third-party solutions. Even if someone uses containerd, we don't know what OCI runtime they prefer. `runc` is common but there are others that are no less valid.

Feedback on this PR should be aimed at helping the author, @PranshuSrivastava, get this ready for merge. For any other discussion, use issues or Slack.

Also:
```
$ dpkg-query --show runc
runc	1.0.1-0ubuntu2~20.04.1
```
(that's a `.deb`)


@PranshuSrivastava the feedback I'd highlight for you here is: https://github.com/kubernetes/website/pull/30882#discussion_r816258989

afbjorklund commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -151,10 +164,11 @@ Install containerd:
-installing the `containerd.io` package can be found at 
-[Install Docker Engine](https://docs.docker.com/engine/install/#server).
+1. Install the `containerd.io` package from the [official containerd website](
+   https://containerd.io/downloads/).Instructions for setting up the Docker
```
You don't need to set up the **Docker** package repository, if installing from the containerd.io.

So either you install the "containerd.io" package from Docker, or you install the "containerd"
package from your distribution (`apt install containerd`) or download it directly from containerd.io

If you want to keep both options (docker.com and containerd.io), they need to be clearly separated.

bart0sh replied:
+1

bart0sh replied:
@afbjorklund thank you.

neolit123 replied:
we should remove the docker repository from the containerd steps and leave the user two options:
- following the guide in https://containerd.io/downloads/ and https://containerd.io/docs/getting-started/#starting-containerd
- use the package manager of the Linux distro

afbjorklund replied:
Would it be possible to supply deb packages for containerd, the same way as for "cri-tools" and "kubernetes-cni" ?

From apt.kubernetes.io, that is.

neolit123 replied:
The package management in k8s needs to be community owned. Currently Google still holds the keys. But also some reorganization of the package repositories might be required for better version flexibility.

The containerd maintainers can open a discussion about that with the k8s SIG Release. In the meantime we should document the best option, IMO.

afbjorklund replied:
Maybe CRI and CNI should be decoupled from the packaging as well, they would be less outdated that way I think.

sftim replied:
@PranshuSrivastava I recommend following the approach outlined in https://github.com/kubernetes/website/pull/30882#discussion_r816261724

@afbjorklund, I'm keen to keep the focus of this PR on just the set of changes we'd make to this page so that the v1.24 docs make sense after the dockershim component is removed. Once this merges, we can consider follow-up PRs that make further improvements.

As the proverb doesn't quite say, the perfect is the enemy of the merged.

afbjorklund replied:
I think changing from docker.com to containerd.io would be enough, go with the tarball if that is all there is.

bart0sh replied:
Can we keep containerd installation instructions as they are as they're not actually related to the purpose of this PR. We can update them in a separate PR I believe.

sftim replied:
After we remove the dockershim, the current instructions are too likely to confuse people. I think fixing that does align with the intent of this PR.

reylejano commented:
/remove-language es id it pt ru uk zh

bart0sh commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -29,6 +29,19 @@ Kubernetes, on Linux:
+/blog/2020/12/08/kubernetes-1-20-release-announcement/#dockershim-deprecation)
+You can check out this [documentation](
+/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you/)
+to understand how this deprecation might affect you. For migrating from
```
Leading "to migrate from dockershim" looks redundant/tautological to me:

For migrating from dockershim ... **to migrate from dockershim.**

PranshuSrivastava replied:
done @bart0sh

bart0sh replied:
@PranshuSrivastava thanks, looks better now.

nate-double-u commented:
/milestone 1.24

PranshuSrivastava commented:
@vincepri PTAL

sftim commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -29,6 +29,19 @@ Kubernetes, on Linux:
 - [Mirantis Container Runtime](#mcr)
 
+{{< note >}}
+Dockershim, the portion of code in Kubernetes that provided direct
+integration with Docker in prior releases, was removed from Kubernetes
+version 1.24. This removal was announced as a [deprecation in Kubernetes v 1.20](
+/blog/2020/12/08/kubernetes-1-20-release-announcement/#dockershim-deprecation)
+You can check out this [documentation](
+/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you/)
+to understand how this deprecation might affect you. To migrate from
+dockershim you can follow [this migration guide](
+/docs/tasks/administer-cluster/migrating-from-dockershim/)
+to migrate from dockershim.
```
```suggestion
Kubernetes releases before v1.24 included a direct integration with Docker Engine,
using a component named _dockershim_. That special direct integration is no longer
part of Kubernetes (this removal was
[announced](/blog/2020/12/08/kubernetes-1-20-release-announcement/#dockershim-deprecation)
as part of the v1.20 release).

You can read
[Check whether Dockershim deprecation affects you](/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you/) to understand how this removal might
affect you. To learn about migrating from using dockershim, see
[Migrating from dockershim](/docs/tasks/administer-cluster/migrating-from-dockershim/).
```

mikebrow replied:
SGTM

sftim commented on content/en/docs/setup/production-environment/container-runtimes.md:
```diff
@@ -428,7 +440,10 @@ You can also find third-party adapters that let you use Docker Engine with Kuber
 through the supported {{< glossary_tooltip term_id="cri" text="Container Runtime Interface">}}
 (CRI).
 
-The following CRI adaptors are designed to work with Docker Engine:
+{{< note >}}
+`overlay2` is the preferred storage driver for systems running Linux kernel version 4.0 or higher,
+or RHEL or CentOS using version 3.10.0-514 and above.
+{{< /note >}}
 
 - [`cri-dockerd`](https://github.com/Mirantis/cri-dockerd) from Mirantis
```
```suggestion
You can also find third-party adapters that let you use Docker Engine with Kubernetes
through the supported {{< glossary_tooltip term_id="cri" text="Container Runtime Interface">}}
(CRI).

The following CRI adapters are designed to integrate Docker Engine with CRI clients (such
as Kubernetes v{{< skew currentVersion >}}):

- [`cri-dockerd`](https://github.com/Mirantis/cri-dockerd) from Mirantis

{{< note >}}
If you use Docker Engine on a node that is running Linux kernel version 4.0 or higher, or that
runs RHEL or CentOS and uses the Linux kernel version 3.10.0-514 and above,
the Kubernetes project suggests using the `overlay2` storage driver.
{{< /note >}}
```

mikebrow replied:
SGTM

sftim reviewed (changes requested):
It looks like this needs some more changes - I've added inline feedback.

sftim commented:
/retitle Revise “Container runtimes” page in light of dockershim removal

bradtopol commented:
Assigning @sftim as the reviewer since he is the one doing the in depth review here.
/unassign
/assign @sftim

sftim commented:
@PranshuSrivastava would you have time to revise this further?

Another option that we have is to take your existing work and build upon it, merging a different PR that is based on what's here.

reylejano commented:
@PranshuSrivastava , are you able to continue to review the comments and make revisions to progress this PR?

If not, I agree with @sftim 's [comment](https://github.com/kubernetes/website/pull/30882#issuecomment-1077867090)

PranshuSrivastava commented:
@reylejano @sftim Yes I have time to work further on this PR. I have addressed all the comments so far and will continue to work on the reviews provided in the following days.
